code/renderercommon/tr_image_png.c

```diff
@@ -2236,3 +2244,281 @@ void R_LoadPNG(const char *name, byte **pic, int *width, int *height) {
+/*
+ * Write PNG chuck header to buffer.
+ */
+static void WriteChunkHeader(void **buffer, void **crcPtr, PNG_ChunkCRC *CRC, int type, int length) {
+	struct PNG_ChunkHeader CH;
+
+	/*
+	 *  Write chuck header
+	 */
+
+	CH.Type = BigLong(type);
+	CH.Length = BigLong(length);
+
+	WriteToBuffer(buffer, &CH, PNG_ChunkHeader_Size);
+
+	/*
+	 *  Init CRC
+	 */
+
+	*CRC = crc32(0, Z_NULL, 0);
+	*CRC = crc32(*CRC, *(byte **)buffer - 4, 4);
+	*crcPtr = *buffer;
+}
+
+/*
+ * Write CRC to buffer.
+ */
+static void WriteCRC(void **buffer, void **crcPtr, PNG_ChunkCRC CRC) {
+	int size = (intptr_t)*buffer - (intptr_t)*crcPtr;
+
+	/*
+	 *  Update CRC
+	 */
+	if (size > 0)
+		CRC = crc32(CRC, *crcPtr, size);
+
+	/*
+	 *  Write CRC
+	 */
+	CRC = BigLong(CRC);
+	WriteToBuffer(buffer, &CRC, PNG_ChunkCRC_Size);
+}
+
+void RE_SavePNG(const char *filename, int width, int height, byte *data, int padding);
+/*
+ * The PNG saver
+ */
+void RE_SavePNG(const char *filename, int width, int height, byte *data, int padding) {
+	void *pngData;
+	size_t pngSize;
+	void *buffer;
+	struct PNG_Chunk_IHDR IHDR;
+	PNG_ChunkCRC CRC;
+	void *crcPtr;
+	int numtEXt = 0;
+#define NUMTEXT 7
+	struct {
+		char key[80];	// PNG limits to 79+'\0'.
+		char text[256]; // PNG allows any length.
+	} tEXt[NUMTEXT];
+	int i;
+	uint8_t *imageData;
+	uint32_t imageLength;
+	Bytef *compressedData = NULL;
+	uLongf compressedDataLength;
+
+	/*
+	 *  Create the png image data from buffer.
+	 */
+
+	imageLength = (width * PNG_NumColourComponents_True + 1) * height;
+	imageData = ri.Malloc(imageLength);
+
+	EncodeImageNonInterlaced8True(width, height, data, 3, padding, imageData);
+
+	/*
+	 *  Compress the png image data.
+	 */
+
+	compressedDataLength = imageLength * 1.01f + 12;
+	compressedData = ri.Malloc(compressedDataLength);
+
+	if (compress(compressedData, &compressedDataLength, imageData, imageLength) != Z_OK) {
+		ri.Free(compressedData);
+		ri.Free(imageData);
+		ri.Printf(PRINT_WARNING, "RE_SavePNG: Failed to compress image data.\n");
+		return;
+	}
+
+	/*
+	 *  Setup text data.
+	 */
+
+	Q_strncpyz(tEXt[numtEXt].key, "Title", sizeof(tEXt[numtEXt].key));
+	ri.Cvar_VariableStringBuffer("version", tEXt[numtEXt].text, sizeof(tEXt[numtEXt].text));
+	numtEXt++;
+	//	Q_strncpyz(tEXt[numtEXt].key, "Author", sizeof (tEXt[numtEXt].key));
+	//	ri.Cvar_VariableStringBuffer("username", tEXt[numtEXt].text, sizeof (tEXt[numtEXt].text));
+	//	numtEXt++;
+	Q_strncpyz(tEXt[numtEXt].key, "Description", sizeof(tEXt[numtEXt].key));
+	ri.CL_GetModDescription(tEXt[numtEXt].text, sizeof(tEXt[numtEXt].text));
+	numtEXt++;
+	Q_strncpyz(tEXt[numtEXt].key, "Mapname", sizeof(tEXt[numtEXt].key));
+	ri.Cvar_VariableStringBuffer("mapname", tEXt[numtEXt].text, sizeof(tEXt[numtEXt].text));
+	numtEXt++;
+	Q_strncpyz(tEXt[numtEXt].key, "Maptitle", sizeof(tEXt[numtEXt].key));
+	ri.CL_GetMapTitle(tEXt[numtEXt].text, sizeof(tEXt[numtEXt].text));
+	numtEXt++;
+	Q_strncpyz(tEXt[numtEXt].key, "Playername", sizeof(tEXt[numtEXt].key));
```

Comment:
not sure if the playername should really get added to the metainfo. the player usually doesn't know this.